File: basic_dsa_three_hundred_nine.java
/**
 * Basic DSA Problems - Missing Number
 * Find the missing number in range [0, n]
 */
public class basic_dsa_three_hundred_nine {

    // Problem: Array nums of n distinct numbers in [0, n]. Return the only number missing.

    /**
     * Sum formula: 0+1+...+n = n*(n+1)/2. Subtract sum(nums) to get missing.
     * Time O(n), Space O(1)
     */
    public static int missingNumber(int[] nums) {
        int n = nums.length;
        int sum = n * (n + 1) / 2;
        for (int x : nums) sum -= x;
        return sum;
    }

    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("Problem: Missing Number");
        System.out.println("=".repeat(60));

        System.out.println("[3,0,1] -> " + missingNumber(new int[]{3, 0, 1}));
        System.out.println("[0,1] -> " + missingNumber(new int[]{0, 1}));
        System.out.println("[9,6,4,2,3,5,7,0,1] -> " + missingNumber(new int[]{9, 6, 4, 2, 3, 5, 7, 0, 1}));
    }
}

File: basic_dsa_three_hundred_nine.cpp
/**
 * Basic DSA - Sum of Two Integers (LeetCode 371)
 * Return a + b without using + or -.
 */
#include <iostream>

int getSum(int a, int b) {
    while (b != 0) {
        unsigned carry = (unsigned)(a & b) << 1;
        a ^= b;
        b = (int)carry;
    }
    return a;
}

int main() {
    std::cout << std::string(60, '=') << "\n";
    std::cout << "Problem: Sum of Two Integers\n";
    std::cout << std::string(60, '=') << "\n";

    std::cout << "1 + 2 -> " << getSum(1, 2) << "\n";
    std::cout << "2 + 3 -> " << getSum(2, 3) << "\n";
}
